File: c++/300. Longest Increasing Subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
<<<<<<< HEAD
        vector<int> dp;
        int n = nums.size();
        for (int i = 0; i < nums.size(); i++) {
            if (dp.empty() || dp.back() < nums[i]) {
                dp.push_back(nums[i]);
                continue;
            }
            int j = lower_bound(dp.begin(), dp.end(), nums[i]) -dp.begin();
            dp[j] = nums[i];
        }
        return dp.size();
    }
};

// dp存储了当前最长上升子序列，遍历数组的时候，
// 将元素插入到dp中合适的位置，遍历完数组之后，dp即是当前最长上升子序列
=======
        vector<int> seq;
        int n = nums.size();
        if (n < 2) {
            return n;
        }
        seq.push_back(nums[0]);
        
        int i;
        int j;
        for (i = 1; i < n; ++i) {
            if (nums[i] > seq.back()) {
                seq.push_back(nums[i]);
                continue;
            }
            j = lower_bound(seq.begin(), seq.end(), nums[i]) - seq.begin();
            seq[j] = nums[i];
        }
        
        int res = seq.size();
        seq.clear();
        return res;
    }
};
>>>>>>> 837bde34d06bb470ffcea088785f591f742870d7

File: python/300. Longest Increasing Subsequence.py
#!usr/bin/env python  
#-*- coding:utf-8 _*-  

class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        res = 0
        if not nums: return res
        n = len(nums)
        dp = [1]*(n+1) # dp表示以i结尾的最长子序列大小
        dp[0] = 0
        for i in range(1, n+1):
            for j in range(i):
                if nums[i-1] > nums[j-1]:
                    dp[i] = max(dp[i], dp[j] + 1)
        return max(dp)

if __name__ == '__main__':
    nums = [1,3,6,7,9,4,10,5,6]
    su =Solution()
    print su.lengthOfLIS(nums)                    